### DISCUSSION TITLE
MQTT Veröffentlichungsintervall

### DISCUSSION THREAD

**Discussion by hoschiking** [hoylabs/OpenDTU-OnBattery#1443] [Q&A] [answered]
Unter MQTT lässt sich das Veröffentlichungsintervall sekundengenau einstellen, leider jedoch nicht auf kleiner als 5 Sekunden.
Das macht einige Regelungen sehr träge, diese neigen dann sogar zum Schwingen, weil die Daten nur alle 5 Sekunden published werden.
Gibt es einen Weg (über ein Config-File oder ähnliches) die Intervalle zu verkürzen?
Ich würd es sogar für sinnvoll halten, wenn man die Einschränkung etwas mildern könnte und ein kürzeres Intervall möglich wäre.
Mit einem Wert ab 2 bis 3 Sekunden wären wohl die meisten Szenarien abgedeckt.

**Comment by spcqike** [comment, marked as answer]
hast du  denn schon einmal in die config geguckt?

![grafik](https://github.com/user-attachments/assets/b776d010-a8cb-4d0c-9449-31888879ae4a)


theoretisch solltest du den Wert dort ändern können. Probiere es einfach mal aus 😄 

Ob das Ganze dann noch stabil läuft, kann ich nicht sagen.

Welche Regelung eigentlich? die openDTU sollte die WR doch regeln.

**Reply by hoschiking** [reply]
Natürlich nicht, wie peinlich. Habs testweise auf 4 Sekunden gestellt, fuktioniert.
Vielen Dank !!!
Ich benötige die Werte für eigene AC-Laderegelung, die läuft im ioBroker.
Dafür habe ich noch einen separaten ESP32 mit OpenDTUoB um 3 MPTT in ioBroker einzubinden,
ebenso noch 2 weitere HM600 von 2 Balkonkraftwerken (hier reicht mir ein 30 Sekunden Intervall völlig)
Ich "missbrauche" also den separaten ESP32 mit OpenDTUoB als MQTT-Datenschleuder für 5 Geräte :-)
Das Projekt ist mittlerweile so vielseitig einsetzbar, es macht echt Spaß damit zu arbeiten!!!

**Comment by hoschiking** [comment]
BTW:
Die Daten vom Stromzähler bekomme ich nur alle 4 Sekunden, vom MPPT nur alle 5 Sekunden.
Gut regeln lässt sich sowas nur, wenn alle Parameter synchron abfragt werden.
In meinem Fall kommt durch den zeitliche Versatz die Regelung immer wieder aus dem Takt.
Siehe Diagramm um ca 12.10 Uhr. Bis auf dieses unschöne Verhalten funktioniert die Regelung schon sehr gut.
![2024-12-04 13_28_43-Window](https://github.com/user-attachments/assets/dc118674-629f-4b47-822f-fad45195b7d9)

**Comment by hoschiking** [comment]
Für Intressierte, falls ähnliche Problematik besteht.
Die Nulleinspeisung über die OpenDUoB zeigte ähnlichs Verhalten, ich habe es zwar stark minimieren können, ganz weg war es nicht.
Erst als ich auf die Idee kam, alle Paramtere möglichst zum gleichen Zeitpunkt zu erfassen, was gar nicht so einfach war,
ist das Schwingen weg. Dies gelang mir auch erst, nachdem ich die Asulesung des Stromzählers von HTTP/JSON auf MQTT umgestellt habe.
Siehe Scrennshots vorher / nacher Beispiel:
![2024-12-04 13_37_36-HM-300 Speicher - Dashboards - Grafana – Mozilla Firefox](https://github.com/user-attachments/assets/23eb6def-ce32-4a81-9587-a95a7963e2e2)

![2024-12-04 13_38_25-HM-300 Speicher - Dashboards - Grafana – Mozilla Firefox](https://github.com/user-attachments/assets/6d895ead-532e-457f-bcfa-708a5be25eca)

**Reply by Snusme83** [reply]
@hoschiking:
Hast du zufällig am Stromzähler nen Volkszähler?
Falls ja, welche Einstellungen braucht es da im Tasmota bzw. in der oDTUoB?

**Reply by hoschiking** [reply]
Leider nein, ich habe einen einfachen selbstgebauten Optokopf, bestehend aus einer Fotodiode und 2 Widerständen an einem esp8266 mit Tasmota.

**Reply by spcqike** [reply]
Tasmota ist Tasmota. Die Hardware, ob gekauft oder gebaut, ist erstmal egal. 

Welche Einstellungen genau brauchst du denn? Tasmota hat viele Beispiele für SML Zähler. Und auch hier im Wiki gibt es viele Hinweise.

**Reply by Snusme83** [reply]
Verwende den Volkszähler aktuell per http+json.
![IMG_3167](https://github.com/user-attachments/assets/813b9088-30ce-4385-8691-0b85b8e8c5fa)
Würde aber gerne per MQTT, da schneller, oder?
Im iOBroker zeigt es mir die Gesamtleistung des Stromzählers an.
Weiß nicht, was ich da eingeben muss beim Stromzähler, wenn ich MQTT verwende.
![IMG_3167](https://github.com/user-attachments/assets/813b9088-30ce-4385-8691-0b85b8e8c5fa)

**Reply by spcqike** [reply]
https://github.com/hoylabs/OpenDTU-OnBattery/wiki/Power-Meter#examples-for-mqtt

**Reply by Snusme83** [reply]
Und weiter? Wo steht da was von einem Volkszähler?
Da geht es um Shellies und tibber,...

**Reply by Snusme83** [reply]
Ich hab keine Wattwerte von den einzelnen Phasen wie es vermutlich ein Shelly hat.
![IMG_3170](https://github.com/user-attachments/assets/ca885018-1bcb-470e-835f-0b051e6b5764)

**Reply by spcqike** [reply]
Zu tasmotas steht da mehr als genug. Ich weiß nicht genau was dir fehlt. 

Wer auch immer sich den Namen  „Volkszähler“ (das ist eigentlich eine ganz andere Software) ausgedacht hat, es ist ein esp mit tasmota.

**Reply by Snusme83** [reply]
![IMG_3171](https://github.com/user-attachments/assets/5b4bceb2-315e-4b0e-9cb1-7dc6cda1412e)

Das MQTT-Topic bräuchte ich. Hab jetzt schon mehrere versucht🤷🏻‍♂️

**Reply by spcqike** [reply]
Weil topic ist halt falsch. Das ist nicht mal ein gültiges topic. 

Wo hast du denn nach deinem topic mal gesucht? Wird dir das im ioBroker angezeigt? Hast du mal mit MQTT Explorer geguckt? Oder auch mal bei tasmota direkt im Wiki gelesen?

**Reply by Snusme83** [reply]
![MQTT Explorer](https://github.com/user-attachments/assets/a6676b0a-738f-4c27-a464-3c7dc81a243a)
Wie gesagt, ich habe schon mehrere Topics probiert, dass das auf dem Bild gezeigte nicht korrekt ist, hab ich mir schon gedacht, weil die oDTUoB mir keinen Stromwert anzeigt.

Laut mqtt Explorer wäre das Topic tele/tasmota_6FF5D0/SENSOR

**Reply by hoschiking** [reply]
Ich sehe gerade, du verwendest den Sonoff Adapter, mit dem wird es nicht funktionieren, da es kein mqtt Broker ist. In die Falle bin ich auch anfangs getappt. Der Sonoff Adapter ist nur ein besserer Parser für Tasmota, mit dem können sich aber keine clients verbinden und können keine Topics abonnieren. Das war der Grund, warum ich lange Zeit den Stromzähler nicht per mqtt abfragen konnte.
Es gibt 2 Lösungen: 
Du musst in ioBroker den mqtt Adapter nutzen, dann hast du aber von allen Tasmota Geräten das mqtt Telegramm nur als json String und musst nachträglich den noch parsen und den den richtigen Datenpunkt herausfischen, oder du kannst weiterhin den Sonoff Adapter nutzen und baust dir per Scripty oder ähnliches eine Routine, die dir zyklisch den Wert in einen Datenpunkt im mqtt Broker schreibt. Den Datenpunkt musst du manuell erzeugen. Letzteres ist meine Lösung, da ich die mit Abstand am besten funktioniert. Bis ich dahin kam, habe ich bestimmt halbes Jahr experimentiert. Im Netz findet man tausende Anleitungen, aber eine, die genau diesen Sachverhalt erklärt und eine Lösung aufzeigt konnte ich ich nirgends finden. Habe damit einige Nächte verbracht, bin aber auch ein Softwarelegasteniker, das ist nicht mein Ding

**Reply by Snusme83** [reply]
![IMG_3170](https://github.com/user-attachments/assets/0d9f0b16-c577-4808-a6aa-35d53c9c6b1a)

Ist das nicht Tasmota?🤔

**Reply by hoschiking** [reply]
Ja klar, steht doch drauf ;-)

**Reply by spcqike** [reply]
Ja das ist tasmota 

> Laut mqtt Explorer wäre das Topic tele/tasmota_6FF5D0/SENSOR

das sieht auch stimmig aus. Welchen JSON Pfad nutzt du da dann? 

> Ich sehe gerade, du verwendest den Sonoff Adapter, mit dem wird es nicht funktionieren, da es kein mqtt Broker ist. In die Falle bin ich auch anfangs getappt. Der Sonoff Adapter ist nur ein besserer Parser für Tasmota, mit dem können sich aber keine clients verbinden und können keine Topics abonnieren.

wie funktioniert der MQTT Explorer dann? Abonniert der nicht alles?

**Reply by Snusme83** [reply]
Sonoff-Adapter oder Tasmota? Was jetzt? Da soll sich einer auskennen!!!

**Reply by Snusme83** [reply]
JSON Pfad bei oDTUoB Stromzähler als Http(s) + JSON: http://192.168.178.126/cm?cmnd=STATUS+8

Dieser funktioniert auch.

**Reply by spcqike** [reply]
Nein. Wenn du MQTT einstellst. Dort brauchst du das topic und kannst optional die Zeile drunter ein json Pfad eingeben. 

Du erhältst auf dem topic doch ein json object.

**Reply by Snusme83** [reply]
JSON Pfad bei oDTUoB Stromzähler als 
- Http(s) + JSON: http://192.168.178.126/cm?cmnd=STATUS+8 <--- funktioniert
- mqtt:  tele/tasmota_6FF5D0/SENSOR <-- Dieses Topic habe ich schon in jeder erdenklichen Form versucht, aber der Stromzähler wird nicht angezeigt!

Bringt es mir überhaupt was bei der oDTUoB, wenn ich anstatt Http(s) + JSON ---- MQTT verwende?

Wir drehen uns irgendwie im Kreis. MQTT wird bei mir ned funktionieren, weil ich das KORREKTE Topic nicht weiß, welches ich eintragen muss!

**Reply by hoschiking** [reply]
Ich habe es dir erklärt, warum es nicht funktionieren wird. Es wäre die geholfen, wenn du auf diesen Hinweis eingehen würdest.
Details siehe auch: 
https://github.com/ioBroker/ioBroker.sonoff

Ich kann dir nur von meinen Erfahrungen berichten.

**Reply by spcqike** [reply]
> Bringt es mir überhaupt was bei der oDTUoB, wenn ich anstatt Http(s) + JSON ---- MQTT verwende?

Ich denke schon. Wenn dein tasmota entsprechend eingestellt ist hast du schneller den Stromzähler wert, und das auch noch Ressourcen schonender auf dem esp (zumindest glaube ich, dass MQTT an der Stelle sparsamer ist, als der http Request) 

> Wir drehen uns irgendwie im Kreis. MQTT wird bei mir ned funktionieren, weil ich das KORREKTE Topic nicht weiß, welches ich eintragen muss!

Ich denke das korrekte topic hast du bereits. Allerdings fehlt dir der korrekte json dazu. Zumindest hast du meine Frage diesbezüglich noch nicht beantwortet. 

@hoschiking 
Kann man sich beim Sonoff Adapter überhaupt mit dem MQTT Explorer anmelden und die topic mit lesen? Das wäre doch bereits eine subscription. Auf „#“

**Reply by hoschiking** [reply]
Die Frage ist doch, ob er auch einen mqtt Broker im Netz am laufen hat und auf welchen Port der eingestellt ist. 
In diesem Screenshot lese ich war vom Sonoff und mit Sonoff funktioniert es nicht. 
![Screenshot_20241204_205700_Firefox](https://github.com/user-attachments/assets/ab74a8ef-0e27-4663-8f9d-83dda4f420d1)

Es sei denn, er hat beides am laufen, Sonoff und mqtt Broker, das ist auch möglich, dann müssen die nur auf andere Ports konfiguriert werden.

**Reply by hoschiking** [reply]
@spcqike 
Meines Wissens nach nicht, zumindest hat es bei mir nie funktioniert und gemäß der Doku im Netz auch nicht. Sonoff ist kein mqtt Broker. Siehe im Link die Tabelle mit den Bemerkungen.

**Reply by spcqike** [reply]
Er kommt doch aber drauf 

https://github.com/hoylabs/OpenDTU-OnBattery/discussions/1443#discussioncomment-11465079

im Screen ist nur einer seiner vielen falschen Versuche. 

Ich will meine mögliche Lösung nicht einfach direkt hinschreiben. Darum halt meine vielen blöden Fragen 😄 ich denke es fehlt mir der json zum topic. 🤷‍♂️ da der MQTT Explorer geht. Und es auch anzeigt.

**Reply by Snusme83** [reply]
Das mit dem eingekreisten Schrott bei MQTT ist doch nur ein schlechtes Beispiel meiner zahlreichen Versuche einen brauchbaren MQTT Topic einzugeben!

Brauche ich denn zu dem vermeintlich korrekten MQTT Topic (tele/tasmota_6FF5D0/SENSOR) noch einen JSON-Pfad? Verstehe das so, dass das optional wäre!
Da hab ich jetzt zum Schluss gar nix mehr eingetragen!

**Reply by hoschiking** [reply]
@spcqike 
Du wirst sicherlich Recht haben, ich kenne mich mit der Materie nicht so gut aus, bis ich das richtige topic herausgefunden habe, habe ich auch Tage gebraucht :-)

**Reply by spcqike** [reply]
> Brauche ich denn zu dem vermeintlich korrekten MQTT Topic (tele/tasmota_6FF5D0/SENSOR) noch einen JSON-Pfad?

gucke dir dein payload zum topic im MQTT Explorer doch einmal genau an. (:

**Reply by hoschiking** [reply]
@Snusme83 
Ja OK, es war einer von vielen Versuchen, dennoch ist es kein Zufall, dass im mqtt Pfad der Name Sonoff vorkommt. 
Soviel ich weiß:
Der json String ist optional, da es davon abhängig ist, in welchen Format dein Tasmota Gerät das Telegramm sendet, als einzelne Objekte oder als json String. Falls als Json String, davon gehe ich aus, musst du noch den json Pfad angeben, damit der korrekte Datenpunkt im json String bekannt ist. Nur dann kann aus dem String der korrekte Datenpunkt gelesen werden. Dabei weiß ich nicht Mal, ob Open DTU es überhaupt kann, da bin ich überfragt.
Gehe doch Mal in die Konsole und kopiere uns Mal so ein Telegramm

**Reply by hoschiking** [reply]
@spcqike 
Vielleicht ist es einfach nur "curr_w"
Wer weiß das schon;-)

**Reply by spcqike** [reply]
> Vielleicht ist es einfach nur "curr_w"

Ich glaube nicht. Etwas mehr müsste es sein.

**Reply by spcqike** [reply]
@Snusme83 
hast du es gelöst bekommen?

Falls nicht, guck dir deine Daten im MQTT Explorer noch einmal an
https://github.com/hoylabs/OpenDTU-OnBattery/discussions/1443#discussioncomment-11465079

den Topic hast du ja ausgewählt, rechts mittig siehst du den Payload. Dieser ist ein JSON Objekt. Also "das Gleiche", was du auch beim HTTP(S)+JSON bekommst. nur, ist der Inhalt bei MQTT anders, als beim HTTP Request

![grafik](https://github.com/user-attachments/assets/2d4809f0-2a27-4ccc-a71a-ae0bfda85f58)

![grafik](https://github.com/user-attachments/assets/4cf5cd16-f694-4a36-9570-5179d04f3951)

Finde den Fehler / Unterschied 😄

**Reply by Snusme83** [reply]
@spcqike: Nein, hab es leider nicht gelöst bekommen, da ich mir im iobroker auch noch was zerschossen hab mit dieser Aktion! Jetzt funktioniert mein Dashboard in Grafana auch nicht mehr!
Aber scheinbar gehört in den Topic "tele/tasmota_6FF5D0/SENSOR" eingetragen plus zusätzlich in optional http JSON Pfad "Time/SML/curr_w"
wenn ich euch richtig verstanden hab!?!

**Reply by spcqike** [reply]
oh, das ist nicht gut. Was hast du denn im ioBroker umgestellt bei der ganzen Sache? und, warum? 🤔 

nein, der JSON Pfad sollte nur ``SML/curr_w`` sein. die ``Time`` und ``SML`` liegen "nebeneinander" vor. ``curr_w`` hingegen ist ein "Unterelement" von ``SML``, zu erkennen an der Einrückung und den ``{}``.

**Reply by Snusme83** [reply]
Wenn ich das nur wüsste! Auf jeden Fall hab ich den Sonoff-Adapter gelöscht im ioBroker!
und jetzt bekomme ich keinen Wert mehr bei curr_w!

![MQTT Explorer2](https://github.com/user-attachments/assets/8f088086-562a-445b-b11b-0fcca0b5c984)

**Reply by Snusme83** [reply]
Jetzt hab ich es hinbekommen. In der oDTUoB kann ich den Stromzähler per mqtt auslesen mit folgenden Einstellungen:

![mqtt Einstellung in oDTUoB](https://github.com/user-attachments/assets/77240595-92f5-4df3-9e97-ee3f18945064)

Warum hat es nicht funktioniert? Ich habe vor lauter Verzweiflung im Tasmota vom Stromzähler den Port von 1883 auf 1885 geändert. Das wusste ich nicht mehr. Jetzt hab ich zurückgestellt auf 1883 - siehe da ein Wert.
Zudem habe ich jetzt für MQTT Topic, sowie Optional: JSON-Pfad die richtigen Einträge!

**Reply by Snusme83** [reply]
oDTUoB bekommt jetzt Daten vom Stromzähler per mqtt!
Jetzt habe ich das Problem, dass ich die momentane Leistung in Grafana nicht mehr anzeigen lassen kann, da ich diesen Wert nirgendswo finde. Vorher hat die oDTUoB die Daten vom Stromzähler per https: json bekommen und ich konnte via iobroker die Daten vom Stromzähler in Grafana integrieren. Wieso geht das per mqtt nicht mehr?

**Reply by Snusme83** [reply]
Der Stromzähler wird in oDTUoB per mqtt auf Port 1883 angezeigt.
Jetzt habe ich in ioBroker den Sonoff-Adapter wieder installiert.
Da zeigt es mir dann auf Port 1885 (in Tasmota umgestellt) alle Werte im Detail an.
Kann man in der mqtt- Konfiguration in Tasmota zwei Ports angeben (1883 + 1885) damit oDTUoB mit Daten versorgt wird, sowie meine Grafana Dashboard!?!

**Reply by hoschiking** [reply]
Du hast jetzt im mqtt Broker einen neuen Datenpunkt als Json String, damit kann Grafana nichts anfangen, daher wirst du so die Daten nicht mehr sehen können. Also lag ich mit meiner Vermutung richtig, dass du vorher den Sonoff Adapter benutzt hast. Siehe hierzu meine Erläuterungen oben. Der Sonoff Adapter parst den Json String und erzeugt die die Datenpunkte selbst, das ist nicht mehr der Fall. Du brauchst also einen Parser, der die die einzelnen Datenpunkte erzeugt. Das ist nicht so trivial, es gibt genug mehr oder weniger gute Anleitungen in den Foren dazu. Am einfachsten geht es per Blockly, wenn man wie ich von Javascript nix versteht. Oder du schreibst den Datenpunkt, den jetzt OpenDTU im mqtt Datenbaum erzeugt in die Datenbank und in stellst in grafana auf den neuen Datenpunkte um, dann sind aber alle historischen Werte bis zur Umstellung nicht mehr in grafana sichtbar

**Reply by Snusme83** [reply]
Ok, jetzt kommt etwas Licht ins Dunkle! die oDTUoB erzeugt aber im Datenbaum die benötigten Werte nicht! Ich recherchiere mal zwecks Blockly.

**Reply by hoschiking** [reply]
Natürlich erzeugt OpenDTU den Datenpunkt, zumindest bei mir.
schaue doch einfach unter mqtt/0/deineOpenDTU/powermeter/powertotal
deineOpenDTU natürlich durch en Namen deines Topics ersetzen.
Edit:
Hast recht, das wird nur erzeugt, wenn der Wert HTTP/JSON geholt wird, sonst nicht.

Daher:
Bevor du wieder tagelang verzweifelst, nimmt dieses Beispiel, ist mit Abstand das einfachste was ich im Netz finden konnte.
Im Netz herrscht ein Wettbewerb, wer den kompliziertesten Parser baut, der hat die dicksten Cohones, so einen Eindruck habe ich manchmal.  
![Screenshot_8](https://github.com/user-attachments/assets/6692daa7-aaa3-4dbb-9c7e-9bddab28ffbc)

Ist ein Beispiel für Raumfühler. Wie man die Pfade einsetzt, hast du mittlerweile gelernt ;-)
ObjektID ist der JSON-Strig von Tasmota, unter attribut kommt der JSON-Pfad rein, hinter "Aktualisiere" sind die zuvor manuell erzeugten Datenpunkte, in die der Parser die Daten schreibt. Vielleicht kannst du es probieren, die alten Datenpunkt vom Sonoff-Adapter nehmen. Falls das klappt, dann hast du den gleichen Datenpunkt wie vor der Umstellung und deine historischen Daten gehen nicht verloren.

**Reply by Snusme83** [reply]
Blöde Frage, wie komme ich zu dem, was du im Bild geschickt hast?

**Reply by hoschiking** [reply]
Im io broker den adapter JScript installieren, dann ist Blockly mit dabei und dann bastelst du dir den Scheiß zusammen ;-)

**Reply by Snusme83** [reply]
![Blockly](https://github.com/user-attachments/assets/ebf7fa72-f879-4d10-96a3-811c2c88ed92)
Wie füge ich unten etwas ein?

**Reply by hoschiking** [reply]
ich denke, das ist das falsche Forum hierfür, dafür gibt es genug Anleitungen im Netz. Es gibt auch ioBroker Forum, schaue doch bitte dort nach. Alles vorkauen möchte dir auch nicht ;-)

**Reply by spcqike** [reply]
> Hast recht, das wird nur erzeugt, wenn der Wert HTTP/JSON geholt wird, sonst nicht.

warum sollte die openDTU es auch machen? Die Daten sind doch bereits im MQTT Broker. Ein Device (das tasmota Ding) published sie bereits. Es macht also wenig Sinn, dass openDTU die daten erneut unter einem zweiten topic sendet. 

Stell doch einfach deine Aufzeichnung entsprechend um, und lausche dort ebenfalls auf das topic vom tasmota. 

Wie bekommt ihr die Daten denn in die Datenbank, die grafana nachher benutzt?

**Reply by hoschiking** [reply]
Ich habe es doch gar nicht kritisiert, habe es auch erst jetzt festgestellt. Ich hatte auch eine Zeit lang dem Stromzähler per http abgefragt und da hat OpenDTU einen mqtt topic erzeugt, daher war ich ob dem Glauben. In diesem Fall erzeugt Tasmota auch einen mqqt topic und dann ist es auch doppelt, also sehe ich persönlich da keinen Zusammenhang.
Meine Lösung habe ich schon mehrfach oben dargelegt ;-)

**Reply by Snusme83** [reply]
Hab es hinbekommen mit dem Netzbezug in Grafana:
![Grafana1](https://github.com/user-attachments/assets/388ce9af-4486-485f-bbe7-9b51520013a2)

**Reply by Snusme83** [reply]
![Grafana2](https://github.com/user-attachments/assets/9ad0ef39-c9f5-4d1f-977e-745686ab5ec9)

**Comment by schlimmchen** [comment]
Siehe 6aaa2de6.

**Reply by hoschiking** [reply]
Danke :-)
